Http2ServerSpec.The Http/2 server implementation should support stream for request entity data should close the stream when we receive a RST after we have half-closed ourselves as well

https://jenkins.akka.io:8498/job/akka-http-nightly/166/AKKA_VERSION=default,Node=akka-http-nightly-node,SCALA_VERSION=2.13.2,jdk=AdoptOpenJDK%2011/


Note: this is a detailed failure report. See [this](https://github.com/akka/akka-http/issues?q=is%3Aopen+label%3Afailed+Http2ServerSpec) other failures in `Http2ServerSpec` 

```
sbt.ForkMain$ForkError: java.lang.AssertionError: received 1 excess messages on WarningFilter(None,Left(),false)
	at akka.testkit.EventFilter.intercept(TestEventListener.scala:121)
	at akka.http.impl.engine.http2.Http2ServerSpec$$anon$12.<init>(Http2ServerSpec.scala:317)
	at akka.http.impl.engine.http2.Http2ServerSpec.$anonfun$new$17(Http2ServerSpec.scala:308)
```

--- a/akka-http-core/src/test/scala/akka/http/impl/util/WithLogCapturing.scala
+++ b/akka-http-core/src/test/scala/akka/http/impl/util/WithLogCapturing.scala
@@ -10,13 +10,15 @@ import akka.actor.ActorSystem
 import akka.event.Logging
 import akka.event.Logging._
 import akka.testkit.{ EventFilter, TestEventListener }
-import org.scalatest.{ Outcome, SuiteMixin, TestSuite }
+import org.scalatest.{ Failed, Outcome, SuiteMixin, TestSuite }
 
 /**
  * Mixin this trait to a test to make log lines appear only when the test failed.
  */
 trait WithLogCapturing extends SuiteMixin { this: TestSuite =>
   implicit def system: ActorSystem
+
+  def failOnSevereMessages: Boolean = false
 
   abstract override def withFixture(test: NoArgTest): Outcome = {
     // When filtering just collects events into this var (yeah, it's a hack to do that in a filter).
@@ -38,14 +40,32 @@ trait WithLogCapturing extends SuiteMixin { this: TestSuite =>
       r
     }
 
-    if (!(res.isSucceeded || res.isPending)) {
+    def flushLog(): Unit = {
       println(s"--> [${Console.BLUE}${test.name}${Console.RESET}] Start of log messages of test that [$res]")
       val logger = new StdOutLogger {}
-      withPrefixedOut("| ") { events.reverse.foreach(logger.print) }
+      withPrefixedOut("| ") {
+        events.reverse.foreach { ev =>
+          if (ev.level == WarningLevel) print(Console.YELLOW)
+          else if (ev.level == ErrorLevel) print(Console.RED)
+          logger.print(ev)
+          if (ev.level <= WarningLevel) print(Console.RESET)
+        }
+      }
       println(s"<-- [${Console.BLUE}${test.name}${Console.RESET}] End of log messages of test that [$res]")
     }
 
-    res
+    if (!(res.isSucceeded || res.isPending)) {
+      flushLog()
+      res
+    } else if (failOnSevereMessages && events.exists(_.level <= Logging.WarningLevel)) {
+      val stats = events.groupBy(_.level).view.mapValues(_.size).toMap.withDefaultValue(0)
+      flushLog()
+
+      Failed(new AssertionError(
+        s"No severe log messages should be emitted during test run but got [${stats(Logging.WarningLevel)}] warnings and [${stats(Logging.ErrorLevel)}] errors (see marked lines above)"
+      ))
+    } else res
+
   }
 
   /** Adds a prefix to every line printed out during execution of the thunk. */

--- a/akka-http2-support/src/main/scala/akka/http/impl/engine/http2/FrameLogger.scala
+++ b/akka-http2-support/src/main/scala/akka/http/impl/engine/http2/FrameLogger.scala
@@ -10,8 +10,10 @@ import akka.stream.scaladsl.{ BidiFlow, Flow }
 import akka.util.ByteString
 
 import scala.collection.immutable.Seq
-
 import FrameEvent._
+import akka.event.Logging
+import akka.stream.Attributes
+import akka.stream.Attributes.LogLevels
 
 /**
  * INTERNAL API
@@ -28,6 +30,7 @@ private[http2] object FrameLogger {
     BidiFlow.fromFlows(
       Flow[FrameEvent].log(s"${Console.RED}DOWN${Console.RESET}", FrameLogger.logEvent),
       Flow[FrameEvent].log(s"${Console.GREEN} UP ${Console.RESET}", FrameLogger.logEvent))
+      .addAttributes(Attributes(LogLevels(Logging.DebugLevel, Logging.DebugLevel, Logging.DebugLevel)))
 
   def logEvent(frameEvent: FrameEvent): String = {
     case class LogEntry(

--- a/akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2Multiplexer.scala
+++ b/akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2Multiplexer.scala
@@ -195,7 +195,7 @@ private[http2] trait Http2MultiplexerSupport { logic: GraphStageLogic with Stage
             sub.data.runWith(subIn.sink)(subFusingMaterializer)
           }
         } else {
-          // stream was cancelled before it we got the response stream
+          // stream was cancelled before we got the response
           sub.data.runWith(Sink.cancelled)(subFusingMaterializer)
           outStreams.remove(sub.streamId)
         }

--- a/akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2ServerDemux.scala
+++ b/akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2ServerDemux.scala
@@ -224,8 +224,8 @@ private[http2] class Http2ServerDemux(http2Settings: Http2CommonSettings, initia
         def onPush(): Unit = {
           val sub = grab(substreamIn)
           pull(substreamIn)
-          multiplexer.registerSubStream(sub)
           handleOutgoingCreated(sub)
+          multiplexer.registerSubStream(sub)
         }
       })
 

--- a/akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala
+++ b/akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala
@@ -134,9 +134,15 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with StageLoggi
     def shutdown(): IncomingStreamState
   }
   case object Idle extends IncomingStreamState {
-    def handle(event: StreamFrameEvent): IncomingStreamState = expectIncomingStream(event, HalfClosedRemote, Open)
+    def handle(event: StreamFrameEvent): IncomingStreamState = expectIncomingStream(event, HalfClosedRemote, ReceivingDataFirst)
     override def handleOutgoingCreated(stream: Http2SubStream): IncomingStreamState = SendingData(stream)
     override def shutdown(): IncomingStreamState = Closed
+  }
+  case class ReceivingDataFirst(buffer: IncomingStreamBuffer) extends ReceivingData(HalfClosedRemote) {
+    override def handleOutgoingCreated(stream: Http2SubStream): IncomingStreamState = Open(buffer)
+    override def handleOutgoingEnded(): IncomingStreamState = Closed
+
+    override protected def onReset(streamId: Int): Unit = multiplexer.cancelSubStream(streamId)
   }
   case class SendingData(outgoingStream: Http2SubStream) extends IncomingStreamState {
     override def handle(event: StreamFrameEvent): IncomingStreamState = expectIncomingStream(event, HalfClosedRemote, Open, _.withCorrelationAttributes(outgoingStream.correlationAttributes))
@@ -229,6 +235,10 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with StageLoggi
     override def shutdown(): IncomingStreamState = Closed
   }
   case object Closed extends IncomingStreamState {
+
+    override def handleOutgoingCreated(stream: Http2SubStream): IncomingStreamState = this
+    override def handleOutgoingEnded(): IncomingStreamState = this
+
     def handle(event: StreamFrameEvent): IncomingStreamState = event match {
       // https://http2.github.io/http2-spec/#StreamStates
       // Endpoints MUST ignore WINDOW_UPDATE or RST_STREAM frames received in this state,

--- a/akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala
+++ b/akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala
@@ -14,7 +14,7 @@ import akka.http.impl.engine.http2.Http2Protocol.{ ErrorCode, Flags, FrameType, 
 import akka.http.impl.engine.http2.framing.FrameRenderer
 import akka.http.impl.engine.server.HttpAttributes
 import akka.http.impl.engine.ws.ByteStringSinkProbe
-import akka.http.impl.util.StringRendering
+import akka.http.impl.util.{ AkkaSpecWithMaterializer, LogByteStringTools, StringRendering }
 import akka.http.scaladsl.model._
 import akka.http.scaladsl.model.headers.{ CacheDirectives, RawHeader }
 import akka.http.scaladsl.settings.ServerSettings
@@ -33,11 +33,11 @@ import scala.collection.immutable
 import scala.collection.immutable.VectorBuilder
 import scala.concurrent.{ Await, Future, Promise }
 import scala.concurrent.duration._
-import scala.util.control.NoStackTrace
 import FrameEvent._
-import akka.http.impl.util.AkkaSpecWithMaterializer
+import akka.event.Logging
 import akka.http.scaladsl.Http2
 import akka.stream.Attributes
+import akka.stream.Attributes.LogLevels
 import akka.stream.testkit.scaladsl.StreamTestKit
 import com.github.ghik.silencer.silent
 
@@ -46,6 +46,8 @@ class Http2ServerSpec extends AkkaSpecWithMaterializer("""
     akka.http.server.http2.log-frames = on
   """)
   with WithInPendingUntilFixed with Eventually {
+  override def failOnSevereMessages: Boolean = true
+
   "The Http/2 server implementation" should {
     "support simple round-trips" should {
       abstract class SimpleRequestResponseRoundtripSetup extends TestSetup with RequestResponseProbes {
@@ -307,20 +309,16 @@ class Http2ServerSpec extends AkkaSpecWithMaterializer("""
       // Reproducing https://github.com/akka/akka-http/issues/2957
       "close the stream when we receive a RST after we have half-closed ourselves as well" in new WaitingForRequestData {
         // Client sends the request, but doesn't close the stream yet. This is a bit weird, but it's whet grpcurl does ;)
-        sendHEADERS(streamId = 1, endStream = false, endHeaders = true, encodeRequestHeaders(request))
-        sendDATA(streamId = 1, endStream = false, ByteString(0, 0, 0, 0, 0x10, 0x22, 0x0e) ++ ByteString.fromString("GreeterService"))
+        sendDATA(streamId = TheStreamId, endStream = false, ByteString(0, 0, 0, 0, 0x10, 0x22, 0x0e) ++ ByteString.fromString("GreeterService"))
 
         // We emit a 404 response, half-closing the stream.
-        emitResponse(streamId = 1, HttpResponse(StatusCodes.NotFound))
+        emitResponse(streamId = TheStreamId, HttpResponse(StatusCodes.NotFound))
 
-        // We don't want to see warnings (which we used to see)
-        EventFilter.warning(occurrences = 0).intercept {
-          // The client closes the stream with a protocol error. This is somewhat questionable but it's what grpc-go does
-          sendRST_STREAM(streamId = 1, ErrorCode.PROTOCOL_ERROR)
-          entityDataIn.expectError()
-          // Wait to give the warning (that we hope not to see) time to pop up.
-          Thread.sleep(3000)
-        }
+        // The client closes the stream with a protocol error. This is somewhat questionable but it's what grpc-go does
+        sendRST_STREAM(streamId = TheStreamId, ErrorCode.PROTOCOL_ERROR)
+        entityDataIn.expectError()
+        // Wait to give the warning (that we hope not to see) time to pop up.
+        Thread.sleep(3000)
       }
       "not fail the whole connection when one stream is RST twice" in new WaitingForRequestData {
         sendRST_STREAM(TheStreamId, ErrorCode.STREAM_CLOSED)
@@ -483,8 +481,12 @@ class Http2ServerSpec extends AkkaSpecWithMaterializer("""
         entityDataOut.sendNext(data1)
         expectDATA(TheStreamId, endStream = false, data1)
 
-        entityDataOut.sendError(new RuntimeException with NoStackTrace)
-        expectRST_STREAM(1, ErrorCode.INTERNAL_ERROR)
+        class MyProblem extends RuntimeException
+
+        EventFilter[MyProblem](pattern = "Substream 1 failed with .*", occurrences = 1).intercept {
+          entityDataOut.sendError(new MyProblem)
+          expectRST_STREAM(1, ErrorCode.INTERNAL_ERROR)
+        }
       }
       "fail if advertised content-length doesn't match" in pending
 
@@ -1039,6 +1041,7 @@ class Http2ServerSpec extends AkkaSpecWithMaterializer("""
 
     final def theServer: BidiFlow[HttpResponse, ByteString, ByteString, HttpRequest, NotUsed] =
       modifyServer(Http2Blueprint.serverStack(settings, system.log))
+        .atop(LogByteStringTools.logByteStringBidi("network-plain-text").addAttributes(Attributes(LogLevels(Logging.DebugLevel, Logging.DebugLevel, Logging.DebugLevel))))
 
     handlerFlow
       .join(theServer)

--- a/akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala
+++ b/akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala
@@ -318,7 +318,7 @@ class Http2ServerSpec extends AkkaSpecWithMaterializer("""
         sendRST_STREAM(streamId = TheStreamId, ErrorCode.PROTOCOL_ERROR)
         entityDataIn.expectError()
         // Wait to give the warning (that we hope not to see) time to pop up.
-        Thread.sleep(3000)
+        Thread.sleep(100)
       }
       "not fail the whole connection when one stream is RST twice" in new WaitingForRequestData {
         sendRST_STREAM(TheStreamId, ErrorCode.STREAM_CLOSED)

